ibis/backends/base/sqlglot/__init__.py

```diff
@@ -1,97 +1,246 @@
-
-
-C = ColGen()
-F = FuncGen()
-NULL = sg.exp.NULL
-FALSE = sg.exp.FALSE
-TRUE = sg.exp.TRUE
-STAR = sg.exp.Star()
-
-
-def make_cast(
-    converter: SqlglotType,
-) -> Callable[[sg.exp.Expression, dt.DataType], sg.exp.Cast]:
-    def cast(arg: sg.exp.Expression, to: dt.DataType) -> sg.exp.Cast:
-        return sg.cast(arg, to=converter.from_ibis(to))
+if TYPE_CHECKING:
+    from collections.abc import Iterator
 
-    return cast
+    import ibis.expr.datatypes as dt
+    import ibis.expr.types as ir
+    from ibis.backends.base.sqlglot.compiler import SQLGlotCompiler
+    from ibis.common.typing import SupportsSchema
+
+
+class SQLGlotBackend(BaseBackend):
+    compiler: ClassVar[SQLGlotCompiler]
+    name: ClassVar[str]
+
+    @classmethod
+    def has_operation(cls, operation: type[ops.Value]) -> bool:
+        # singledispatchmethod overrides `__get__` so we can't directly access
+        # the dispatcher
+        dispatcher = cls.compiler.visit_node.register.__self__.dispatcher
+        return dispatcher.dispatch(operation) is not dispatcher.dispatch(object)
+
+    def _transform(
+        self, sql: sge.Expression, table_expr: ir.TableExpr
+    ) -> sge.Expression:
+        return sql
+
+    def table(
+        self, name: str, schema: str | None = None, database: str | None = None
+    ) -> ir.Table:
+        """Construct a table expression.
+
+        Parameters
+        ----------
+        name
+            Table name
+        schema
+            Schema name
+        database
+            Database name
+
+        Returns
+        -------
+        Table
+            Table expression
+        """
+        table_schema = self.get_schema(name, schema=schema, database=database)
+        return ops.DatabaseTable(
+            name,
+            schema=table_schema,
+            source=self,
+            namespace=ops.Namespace(database=database, schema=schema),
+        ).to_expr()
+
+    def _to_sqlglot(
+        self, expr: ir.Expr, limit: str | None = None, params=None, **_: Any
+    ):
+        """Compile an Ibis expression to a sqlglot object."""
+        table_expr = expr.as_table()
+
+        if limit == "default":
+            limit = ibis.options.sql.default_limit
+        if limit is not None:
+            table_expr = table_expr.limit(limit)
+
+        if params is None:
+            params = {}
+
+        sql = self.compiler.translate(table_expr.op(), params=params)
+        assert not isinstance(sql, sge.Subquery)
+
+        if isinstance(sql, sge.Table):
+            sql = sg.select(STAR).from_(sql)
+
+        assert not isinstance(sql, sge.Subquery)
+        return [self._transform(sql, table_expr)]
+
+    def compile(
+        self, expr: ir.Expr, limit: str | None = None, params=None, **kwargs: Any
+    ):
+        """Compile an Ibis expression to a ClickHouse SQL string."""
+        queries = self._to_sqlglot(expr, limit=limit, params=params, **kwargs)
```

Comment:
What is the advantage of having `._to_sqlglot()` separated from `.compile()`? Feels like having a `.compile()` should be sufficient.

Reply:
I mean we don't necessarily need to return with a string here, it could return with the sqlglot expression instead.

Reply:
`_to_sqlglot` can return with a `list` of sqlglot expressions if there are DDL statements like `CREATE TEMPORARY FUNCTION`, which shows up in the BigQuery backend.

I can revert this change here though, and only add it when I port the bigquery backend.

Reply:
Reverted this for duckdb and clickhouse. I'll remove the datafusion version of this method in #7434.

Reply:
Going to address this one in a follow-up. Changing to return a sqlglot object from `compile` breaks a bunch of tests.